---
title: "Class19"
author: "Wenxi Tang"
format: pdf
editor: visual
---

# Web scrapping

Here I extract the CDC figures for Pertussis cases in USA: https://www.cdc.gov/pertussis/surv-reporting/cases-by-year.html

>Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r echo=FALSE}
cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L),
         Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617)
       )
```


Let's make a plot of number of Pertussis cases per year. 

```{r}
library(ggplot2)
```

>Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

The vaccine effectively brought down number of cases. The fluctuation in number of cases may reflect vaccine hesitancy, vaccine side-effects. The newly introduced aP vaccine seems not as effective as the wP vaccines as after the new vaccine was introduced, the number of cases started to fluctuate again. 

```{r}
base <- ggplot(cdc) + aes(Year, Cases) +
          geom_point() +
          geom_line() +
          scale_y_continuous()

base
```

The first vaccine (so called wP) was introduce in the USA in 1946. Let's add this as a line to our plot. 

The US, and many other countries, switched from the original wP vaccine to a new aP vaccine (acellular) in 1996. Let's add this line to our plot. 


```{r}
#add lines for introduction for vaccines. 
base + 
  geom_vline(xintercept = 1946, linetype = 5, col = "blue") +
  geom_vline(xintercept = 1996, linetype = 5, col = "red")
```

>Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

The number of cases started to fluctuate and increase a bit before it was introduced. Probably the aP vaccine is less effective compared to the wP vaccine and requires a booster after a while (explains why the first a few years were still pretty stable). Or it might be the waning of immunity in adolescents originally primed as infants with the newer aP vaccine as compared to the older wP vaccine.

# Explore CMI-PB data

To read these types of files into R we will use the read_json() function from the jsonlite package. Note that if you want to do more advanced querys of APIs directly from R you will likely want to explore the more full featured rjson package. The big advantage of using **jsonlite** for our current purposes is that it can simplify JSON key-value pair arrays into R data frames without much additional effort on our part.

```{r}
library(jsonlite)

subject <- read_json("http://cmi-pb.org/api/subject", simplifyVector = TRUE)

head(subject)
dim(subject)
```

>Q4. How many aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

>Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```


>Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$race, subject$biological_sex)
```

Let's look at the specimen table next: 

```{r}
specimen <- read_json("http://cmi-pb.org/api/specimen", simplifyVector = TRUE)
head(specimen)
dim(specimen)
```

To know whether a given specimen_id comes from an aP or wP individual we need to link (a.k.a. “join” or merge) our specimen and subject data frames. The excellent dplyr package (that we have used previously) has a family of join() functions that can help us with this common task:

>Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
library(dplyr)
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```

>Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
#read the titer dataset
titer <- read_json("http://cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 

#joining the titer dataset with the meta dataset
abdata <- inner_join(titer, meta)
dim(abdata)
head(abdata)
```

> Q. What do you notice about the "visit" number 8? 

It's only 80 compared to others which are all a few thousands. 

```{r}
table(abdata$visit)
```


>Q. How many isotypes do we have in this dataset?


>Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(titer$isotype)
```

>Q12. What do you notice about the number of visit 8 specimens compared to other visits?

# Examine IgG1 Ab titer levels


```{r}
#filter only the igg1 data
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
dim(ig1)
```

>Q. How many antigens are there?

```{r}
table(ig1$antigen)
```

Analysis of the whole dataset: antigen levels - i.e. a plot of antigen vs MFI. 

```{r}
ggplot(ig1) +
  aes(MFI, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2) +
  theme_bw()
```

>Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?

The FIM2/3, TT, PRN increases along with visits. Probably because these antigens are extracellular and easily recognized by antibodies. 

```{r}
ggplot(ig1) +
  aes(MFI, antigen, col = infancy_vac) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
```


> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).

```{r}
#MEASLE ANTIGEN LEVELS
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```
```{r}
#FIM2/3 ANTIGEN LEVEL
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
#PT ANTIGEN LEVEL
filter(ig1, antigen=="PT") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```


>Q16. What do you notice about these two antigens time courses and the FIM2/3 data in particular?

The level of FIM2/3 and PT shows an increasing trend and shows a peak around 4th~6th visits. Definitely an obvious increase compared to the measles control. 

>Q17. Do you see any clear difference in aP vs. wP responses?

They show similar increasing trends along with visits. No clear differences. 

# Obtaining CMI-PB RNASeq data

```{r}
#read the input data
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
head(rna)
```

To facilitate further analysis we need to “join” the rna expression data with our metadata meta, which is itself a join of sample and specimen data. This will allow us to look at this genes TPM expression values over aP/wP status and at different visits (i.e. times):

```{r}
ssrna <- inner_join(rna, meta)
dim(ssrna)
```

>Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```


>Q19.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

It has a maximum level at the 4th visit and starts to drop on the 5th visit. 

>Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?

The expression peak is earlier than the antibody peak, which is around the 5th visit. And the antibody titer stays longer. This makes sense because we need the expression, protein production is followed and it can stay for a while.

We can dig deeper and color and/or facet by infancy_vac status:

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

There is however no obvious wP vs. aP differences here even if we focus in on a particular visit:

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

>Q. Is RNA-Seq expression levels predictive of Ab titers?

Yes, it's quite predictive. 

>Q. What differentiates aP vs. wP primed individuals?

There are some differences between aP and wP individuals but we still not sure how to differentiate them. 


